File: leetcode/problems/310.minimum-height-trees.py
#
# @lc app=leetcode id=310 lang=python3
#
# [310] Minimum Height Trees
#

# @lc code=start


import collections


class Solution:
    def findMinHeightTrees(self, n: int, edges: list[list[int]]) -> list[int]:
        if not edges:
            return [0]

        graph = collections.defaultdict(list)
        for edge in edges:
            u, v = edge[0], edge[1]
            graph[u].append(v)
            graph[v].append(u)

        leaves = []
        for i in range(n):
            if len(graph[i]) == 1:
                leaves.append(i)

        while n > 2:
            n -= len(leaves)

            new_leaves = []
            for leaf in leaves:
                node = graph[leaf].pop()
                graph[node].remove(leaf)

                if len(graph[node]) == 1:
                    new_leaves.append(node)

            leaves = new_leaves

        return leaves


# @lc code=end

File: leetcode/problems/310.minimum-height-trees.java
/*
 * @lc app=leetcode id=310 lang=java
 *
 * [310] Minimum Height Trees
 */

// @lc code=start

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        if (edges.length == 0) {
            ArrayList<Integer> list = new ArrayList<Integer>();
            list.add(0);
            return list;
        }

        HashMap<Integer, HashSet<Integer>> graph = new HashMap<>();
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];

            if (!graph.containsKey(u)) {
                graph.put(u, new HashSet<>());
            }
            if (!graph.containsKey(v)) {
                graph.put(v, new HashSet<>());
            }

            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        HashSet<Integer> leaves = new HashSet();
        for (int i = 0; i < n; i++) {
            if (graph.containsKey(i) && graph.get(i).size() == 1) {
                leaves.add(i);
            }
        }

        while (n > 2) {
            n -= leaves.size();

            HashSet<Integer> new_leaves = new HashSet<>();
            for (int leaf : leaves) {
                int node = graph.get(leaf).iterator().next();
                graph.get(node).remove(leaf);

                if (graph.get(node).size() == 1) {
                    new_leaves.add(node);
                }
            }
            leaves = new_leaves;
        }

        return new ArrayList<Integer>(leaves);
    }
}
// @lc code=end
